### DISCUSSION TITLE
python使用NCNN模型出现精度损失

### DISCUSSION THREAD

**Discussion by TummyWang** [Tencent/ncnn#4842] [Q&A]
我尝试用python的NCNN库验证我通过onnx转化的ncnn模型的有效性，但转化后的ncnn模型效果精度大幅低于原始的onnx模型。
我使用的转化方法是直接利用网站进行转化：https://convertmodel.com。之前用过是有效的。

我尝试使用同一张图片，同时通过onnx和ncnn模型，计算最后输出的相似度，但相似度极低。尝试了很多很多种ncnn图片读取方法，均不不理想，其中一个验证方法如下：
def convert(img_path, net,sess):
    image = cv2.imread(img_path)
    image = cv2.resize(image, (224, 224))
    image = np.transpose(image, (2, 0, 1))
    image1=image.astype(np.float32)
    image = image.astype(np.float32) / 255.0 # 将图像数据转换为 float 类型并进行归一化
    image = np.expand_dims(image, axis=0)
    input_name = sess.get_inputs()[0].name
    output_name = sess.get_outputs()[0].name
    outputs = sess.run([output_name], {input_name: image})
    feature1 = outputs[0][0]
    image = image[0]
    image = cv2.cvtColor(image1, cv2.COLOR_BGR2RGB) #加与不加效果差别均不理想
    mat_in = ncnn.Mat.from_pixels(image, ncnn.Mat.PixelType.PIXEL_RGB,224,224)

    ex = net.create_extractor()
    ex.input("x.1", mat_in)  # replace "input_name" with the actual input name of your model
    mat_out = ncnn.Mat()
    ex.extract("1864", mat_out)  # replace "output_name" with the actual output name of your model
    feature2 = np.array(mat_out, copy=False)
    feature=torch.cosine_similarity(torch.from_numpy(feature1),torch.from_numpy(feature2), dim=0, eps=1e-8)
    return feature

后来我尝试验证np.array(ncnn.Mat(image1))-image1
但返回值并不是0矩阵。似乎与pytorch-numpy之间的任意转化有非常显著的差异。

不知道是python 的ncnn库有问题，还是我使用ncnn加载图片的方式有问题？

**Comment by magicse** [comment]
may be ncnn.Mat.from_pixels work with values 0..255 and not float [use ncnn with opencv](https://github.com/Tencent/ncnn/wiki/use-ncnn-with-opencv) ?

**Reply by TummyWang** [reply]
非常感谢你的解答，错误的原因确实在此，当我首先用 ncnn.Mat.from_pixels，之后再对mat_in进行mat_in.substract_mean_normalize(mat_in,norm=[1/255.0,1/255.0,1/255.0])  误差便消失了。
另外ncnn.Mat.from_pixels(image1, ncnn.Mat.PixelType.PIXEL_RGB, 224, 224)-image1返回结果也是0矩阵
再次表达感谢！
